bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,802 @@
+                    let _ = responder.send(Err(err.into()));
+                }
+            }
+            ApiError::FailedToReconstructRackSecret => {
+                if let Some(responder) = self.rack_secret_responder.take() {
+                    let _ = responder.send(Err(err.into()));
+                }
+            }
+            // Nothing to do for these variants
+            // We already loggged the error at the top of this method
+            ApiError::FailedToDecryptExtraShares
+            | ApiError::UnexpectedResponse { .. }
+            | ApiError::ErrorResponseReceived { .. }
+            | ApiError::InvalidShare { .. } => {}
+        }
+    }
+
+    // Handle messages from connection management tasks
+    async fn on_conn_msg(&mut self, msg: ConnToMainMsg) {
+        let unique_id = msg.handle_unique_id;
+        match msg.msg {
+            ConnToMainMsgInner::ConnectedServer {
+                accepted_addr,
+                addr,
+                peer_id,
+            } => {
+                let Some(accepted_handle) =
+                   self.accepted_connections.remove(&accepted_addr) else
+                {
+                    error!(
+                        self.log,
+                        "Missing AcceptedConnHandle";
+                        "accepted_addr" => accepted_addr.to_string(),
+                        "addr" => addr.to_string(),
+                        "remote_peer_id" => peer_id.to_string()
+                    );
+                    panic!("Missing AcceptedConnHandle");
+                };
+
+                // Ignore the stale message if the unique_id doesn't match what
+                // we have stored.
+                if unique_id != accepted_handle.unique_id {
+                    return;
+                }
+
+                // Gracefully close any old tasks for this peer if they exist
+                self.remove_established_connection(&peer_id).await;
+
+                // Move from `accepted_connections` to `connections`
+                let handle = PeerConnHandle {
+                    handle: accepted_handle.handle,
+                    tx: accepted_handle.tx,
+                    addr,
+                    unique_id: accepted_handle.unique_id,
+                };
+                self.connections.insert(peer_id.clone(), handle);
+                if let Err(e) =
+                    self.fsm.on_connected(Instant::now().into(), peer_id)
+                {
+                    // This can only be a failure to init the rack, so we
+                    // log it as an error and not a warning. It is unrecoverable
+                    // without a rack reset.
+                    error!(self.log, "Error on connection: {e}");
+                } else {
+                    self.deliver_envelopes().await;
+                }
+            }
+            ConnToMainMsgInner::ConnectedClient { addr, peer_id } => {
+                let handle =
```

Comment:
Also, I'm wondering if a misbehaving client could cause us to panic here – if it tries to connect to the same server repeatedly, we'll have removed it from `self.negotiating_connections` after the first connection...

Reply:
Great find! This is definitely a bug. On a connection failure, the connection will be retried inside the loop in `spawn_connection_initiator_task`. But we won't actually have any handle in `self.initiating_connections`, since we only load that once and remove it the first time this message is handled.

I believe we would have found this once I add state persistence and restart tests, as then a reconnection would occur and we would have panicked. But I'm thankful your eagle eye saved me that debugging!

Reply:
Also, the reason the error is in the `ConnectedServer` case is that there is no loop in that task, as the server side task gets recreated when the remote peer connects to us, versus us connecting to a remote peer in a loop.

Reply:
Damnit, this reveals a larger issue for status queries. We never go back to the negotiatitng/initiating_connection state, so it will always report we are established. I'm going to have to take this loop out I think.

Reply:
> Damnit, this reveals a larger issue for status queries. We never go back to the negotiatitng/initiating_connection state, so it will always report we are established. I'm going to have to take this loop out I think.

Actually, no I already can handle this with the `Disconnect` message. I just have to move the handle back into  `self.initiating_connections`

Reply:
Removed an incorrect comment related to the fix: 4291e47ca